[FEATURE] Add CLI Command

jdoubleu opened the pull request:
Adds a command for use with console to precompile and profile fluid templates.

See also https://github.com/NamelessCoder/typo3-cms-fluid-precompiler/issues/2

Commit:
Implement compile cli command

Commit:
Change nofail and failed argument

The 'nofail' argument was renamed to 'fail' to improve understanding. The 'failed' argument no longer exists because it could be confused with the new 'fail' argument and has therefore changed to (show) 'only-failed'.

Commit:
Use base exception when catching compiling errors and failing command

Commit:
Fix php error in 'fail' switch when compiling failed

Commit:
Update command's phpDoc

Commit:
Implement profile command

NamelessCoder commented on Classes/Command/FluidPrecompileCommandController.php:
```diff
@@ -0,0 +1,220 @@
+        try {
+            $result = $this->getPrecompilerService()->warmup($extension);
+        } catch (\Exception $e) {
+            $result = [];
```
This looks like a misunderstanding about what `$fail` actually does.

It does not suppress errors. It makes the warmup fail if any one template could not compile - which when it happens, does not raise an exception. So no try-catch here, but instead, a `throw` down below if there is one or more than failed templates.

jdoubleu replied:
Sorry, I can't quite follow. Exceptions will be thrown at the end of the function if `--fail` was provided.

I understand the `--fail` flag as following:
If there's any kind of an error (e.g. compiling, TYPO3 or "uncompilable templates") **and** the `--fail` flag was provided the program should exit with a non-zero return code. In this case a thrown exception would cause this.
But when the `--fail` flag isn't provided it shouldn't, because the program was told not to fail.

Or would you like to see the application "failing" if an internal error (i.e. the compiling failed) occurs while staying silent for all contextual error (i.e. "uncompilable templates") when no `--fail` flag was provided.

jdoubleu replied:
I've reworked the `--fail` logic:
Exceptions thrown while compiling/warmup won't be catched anymore.
Only if `--faill` was provided **and** there were uncompilable templates, a custom exception will be thrown which causes a non-zero status code.

See https://github.com/NamelessCoder/typo3-cms-fluid-precompiler-module/pull/6/commits/7215c5d13366cdd2d56357e7cb58a0732ec595a1

NamelessCoder reviewed (changes requested):
Everything else looks pretty damn good, the one thing that could improve it is to use strict mode (up to 7.0 so no nullable). Very minor since only a single return type annotation could be added anyway.

The other comment in code is however more important to address!

mbrodala commented on Classes/Command/FluidPrecompileCommandController.php:
```diff
@@ -0,0 +1,220 @@
+/**
+ * Class FluidPrecompileCommandController
+ */
+class FluidPrecompileCommandController extends CommandController
```
I suggest making this a Symfony command instead, this way we skip the whole Extbase overhead. The command can be used only with TYPO3v8 or newer anyways.

NamelessCoder replied:
Good point, let's do that.

jdoubleu replied:
I'd normally agree with you on this.
I prefer a Symfony Command, too, because TYPO3's Command API is very limited.

However I wanted to stay compatible with the TYPO3 Core (or Extbase) and especially the [TYPO3 Console](https://github.com/TYPO3-Console/TYPO3-Console).

Another benefit of the Command Controller is, that it can be used by the scheduler extension.

A Symfony Command would also require some sort of Symfony Application bootstrapping (even if its a oneliner). How would you've done this? Creating an extra _bin_ script?

I'd suggest to add an additional Symfony Command but also keep the TYPO3 Command.

mbrodala replied:
TYPO3 [natively supports](https://docs.typo3.org/typo3cms/extensions/core/8.7/singlehtml/Index.html#feature-73042-introduce-native-support-for-symfony-console) and uses Symfony commands. And starting from TYPO3v9 you have a [scheduler task for executing registered Symfony tasks](https://docs.typo3.org/typo3cms/extensions/core/latest/singlehtml/#feature-79462-introduce-scheduler-task-to-execute-console-command) too.

TL;DR: it's easier than you think. ;-)

jdoubleu replied:
Okay, I didn't know. Sadly, the core documentation on this very poor and as far as I've checked, the extbase commands rely on the TYPO3 Command.

I'm going to migrate the CommandController to a Symfony Command.

jdoubleu commented:
Additional I'd like to add tests. This extension has no tests yet, so I need to setup/configure the testing environment/framework.

Commit:
Add warning to compile command to notify the user about error handlers

Commit:
Migrate fluid precompile CommandController to symfony commands

Commit:
Rework output and fail handling for compile command

There's now always a generic information output (unless --quiet is provided). The templates information table is only shown for verbose output (-v).
Exceptions during compiling will always be thrown.
When --fail is provided and there were uncompilable templates an exception is thrown to return with a non-zero status code.

jdoubleu commented:
I've migrated the CommandController into native Symfony Commands.

One downside is, that these commands are no longer available in the [TYPO3 Console](https://github.com/TYPO3-Console/TYPO3-Console).

One thing I noticed was that this extension already defines some TYPO3 core commands like `backend:lock` (see [`Commands.php`](https://github.com/NamelessCoder/typo3-cms-fluid-precompiler-module/blob/02ec114d94051cb4d1aa9ec88d24385cd85a2ab9/Configuration/Commands.php)) causing an exception for the `typo3` command in TYPO3 CMS projects because this Command was already registered:
```
# vendor/bin/typo3
TYPO3\CMS\Core\Console\CommandNameAlreadyInUseException: Command "backend:lock" registered by "cms_fluid_precompiler_module" is already in use in file vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandRequestHandler.php on line 148
Stack trace:
  1. TYPO3\CMS\Core\Console\CommandNameAlreadyInUseException->() vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandRequestHandler.php:148
  2. TYPO3\CMS\Core\Console\CommandRequestHandler->populateAvailableCommands() vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandRequestHandler.php:72
  3. TYPO3\CMS\Core\Console\CommandRequestHandler->handleRequest() vendor/typo3/cms/typo3/sysext/core/Classes/Core/Bootstrap.php:315
  4. TYPO3\CMS\Core\Core\Bootstrap->handleRequest() vendor/typo3/cms/typo3/sysext/core/Classes/Console/CommandApplication.php:75
  5. TYPO3\CMS\Core\Console\CommandApplication->run() vendor/typo3/cms-cli/typo3:22
  6. {closure}() vendor/typo3/cms-cli/typo3:23
```

Commit:
Implement option to temporary disable error handlers

Commit:
Format output of verbose rendering information in a readable format

mbrodala commented:
@jdoubleu Symfony commands registered for TYPO3 work perfectly fine in the TYPO3 Console, at least in 4.x and 5.x.

And I don't know what the current `Commands.php` is supposed to do. As you can see what it currently does is invalid. @NamelessCoder can you shed some light on this?

jdoubleu commented:
> @jdoubleu Symfony commands registered for TYPO3 work perfectly fine in the TYPO3 Console, at least in 4.x and 5.x.

Okay, even better. Didn't really test it with the TYPO3 Console. Maybe this was just a mistake on my side.

@mbrodala @NamelessCoder could you please review the new changes?

mbrodala commented on Classes/Command/FluidPrecompilerServiceTrait.php:
```diff
@@ -0,0 +1,24 @@
+    {
+        $objectManager = GeneralUtility::makeInstance(ObjectManager::class);
+
+        return $objectManager->get(FluidPrecompilerService::class);
```
There's no need for Extbase here, you can just use `GeneralUtility::makeInstance(FluidPrecompilerService::class)` straight here.

mbrodala commented on Classes/Command/FluidPrecompilerServiceTrait.php:
```diff
@@ -0,0 +1,24 @@
+/**
+ * Trait used by commands to get the FluidPrecompilerService
+ */
+trait FluidPrecompilerServiceTrait
```
I'm not sure we really need this trait; this could also be done with an abstract command which provides this getter.

NamelessCoder replied:
Duplicating the getter in two classes is also perfectly acceptable if the alternative is to create a single-method trait or abstract class. In fact, I'd say duplication is preferred for single-line, single-function cases.

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+/**
+ * A console command for precompiling Fluid templates
+ */
+class FluidTemplatePrecompileCommand extends Command
```
Should be named `FluidPrecompileCommand` since it also warms up partials and layouts, see https://github.com/TYPO3/Fluid/blob/f583bad/src/Core/Cache/StandardCacheWarmer.php#L77-L79

NamelessCoder replied:
Agree!

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+    protected function configure()
+    {
+        $this
+            ->setDescription('Precompile Fluid templates')
```
`... templates, partials and layouts`

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+            ->addArgument(
+                'extension',
+                InputArgument::OPTIONAL,
+                'Extension key of which the fluid template should be compiled, leave empty to compile all',
```
`... the Fluid resources ...` (especially _Fluid_, not _fluid_. ;-) )

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+                'show-only-failed',
+                null,
+                InputOption::VALUE_NONE,
+                'Show only failed fluid templates from compiling. Only in conjunction with -v'
```
`... Fluid templates which failed to compile.`

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+                'limit',
+                'l',
+                InputOption::VALUE_REQUIRED,
+                'Limit of template files of each extension shown. Only in conjunction with -v',
```
`... to show.`

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+                'disable-error-handler',
+                null,
+                InputOption::VALUE_NONE,
+                'Temporary disable any registered error handler for compiling'
```
`Temporarily`

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+            $outputHelper->warning(
+                "\nIt seems like there is an error handler registered."
+                . "\nIn some cases this can prevent the precompiling to run properly."
+                . "\nAn ErrorHandler might raise an exception for a PHP Warning."
```
Here I'd suggest the `--disable-error-handler` flag.

jdoubleu replied:
It already applies (see the if-condition): If this flag is provided the warning won't be displayed.

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+
+        // General information
+        $templatesCount = 0;
+        foreach ($result as $extensionResult) {
```
This could be done with `array_map()` and `array_sum()`

jdoubleu replied:
`array_sum` would use the array's values. But these are arrays too (for each extension an array of templates).

NamelessCoder replied:
What Mathias meant was `$templatesCount = array_sum(array_map('count', $result));` does the same. First it creates an array of counts of results for each extension, then sums those up.

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+
+        $outputHelper->section('General');
+        $outputHelper->text(vsprintf('Compiled %d extension(s) with %d template(s) in total.', [count($result), $templatesCount]));
+        $outputHelper->text(vsprintf('%d template(s) could not be compiled.', [$uncompilable]));
```
This should be shown only if necessary and as warning/error.

jdoubleu replied:
For me this is some general information about the procompiling job.

The command will also fail if the amount of uncompilable templates is greater than 0 and the `--fail` flag was provided.

How about showing this for the first verbosity level? When using `-v` this will be shown.

jdoubleu replied:
I see this as an important information, even when no verbose output is requested and the command should not fail.
You may want to warmup your caches, including the Fluid resources, and have some nice metric.

I'm also against the `-v` option because this also shows the detailed compilation report where you can already extract this metric.

How do you think about this?

NamelessCoder replied:
It's vital information and should be shown. The trick is to only error out if failures are wanted - but the information that you do have uncompilable templates is pretty important, I agree.

mbrodala commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,163 @@
+        }
+
+        if ($fail && $uncompilable > 0) {
+            throw new \Exception(sprintf('Could not compile %d templates.', $uncompilable), 1518205183);
```
Please use `return` instead here.

jdoubleu replied:
Do you mean to remove the exception and return an invalid status code or to return the exception?
Latter would be wrong because the Command's `execute()` function should only return a status code.

I'd prefer a thrown exception because it is more highlighted in a console. A simple non-zero return code might not be as noticeable as a big red box.

NamelessCoder replied:
Failure in this CLI task is by definition supposed to be an error return code, not a runtime exception. The reason behind that is that an exception is indistinguishable from an internal error, whereas a good old return code communicates exactly what it should: that the task failed because you told it to fail if it couldn't compile.

However, a message to go along with that return code certainly wouldn't hurt.

mbrodala commented on Configuration/Commands.php:
```diff
@@ -16,5 +16,11 @@
     'referenceindex:update' => [
         'class' => \TYPO3\CMS\Backend\Command\ReferenceIndexUpdateCommand::class
+    ],
+    'fluidprecompile:compile' => [
```
I'd suggest `fluid:precompile`

jdoubleu replied:
I'm not sure if this is going to conflict with the EXT:fluid.
This extension is name `fluid_precompiler_module` so this should be the commands vendor, shouldn't it?

NamelessCoder replied:
Fluid itself will never provide a `precompile` or `profile` command - it's outside the responsibility of the engine package, so there's no risk in using those. If anything, Fluid will at one time provide a `render` command and that's about it.

mbrodala commented on Configuration/Commands.php:
```diff
@@ -16,5 +16,11 @@
+    'fluidprecompile:compile' => [
+        'class' => \NamelessCoder\CmsFluidPrecompilerModule\Command\FluidTemplatePrecompileCommand::class
+    ],
+    'fluidprecompile:profile' => [
```
Here I'd use `fluid:profile`

jdoubleu replied:
see above.

Commit:
Update command arguments' and options' descriptions

Commit:
Remove unused imports

Commit:
Update FluidPrecompileService instantiating

Drop usage of ObjectManager in favor to avoid extbase bootstrapping.

jdoubleu commented:
Any updates on this?

mbrodala commented:
@NamelessCoder are your requested changes still valid? From my POV the current state works quite well (really checked on CLI), small details like strict types can be added later.

NamelessCoder commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,161 @@
+        $disableErrorHandlers = $input->getOption('disable-error-handler');
+
+        // Test exception handlers
+        if (set_error_handler(function() {}) && !$disableErrorHandlers) {
```
This really should not be necessary - the default behavior is to catch exceptions internally and analyse them. We should not need to replace the error handler: if an exception really is raised, it is an error in the templates or an incompatibility which prevents it from being parsed without errors.

FYI: Fluid has an internal error handling device but it is *not* a PHP registered error handler. It is quite selective about which exceptions are actual errors and which are parsing errors - and only "real" errors would break in a way that triggers whichever error handler is registered. Additionally, PHP warnings must definitely not be suppressed because they indicate real errors (such as not checking input types, abusing parse-time execution, etc.)

Therefore I would rather we do not make it possible to manipulate this low level of the runtime. The right way is to show the errors and expect the user to fix them.

NamelessCoder commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,161 @@
+     * @param mixed $data
+     * @return string
+     */
+    protected static function formatReadable($data)
```
This method does not need to be static - and there is only a single usage. I'd remove it.

NamelessCoder commented on Classes/Command/FluidTemplateProfileCommand.php:
```diff
@@ -0,0 +1,162 @@
+                $rows
+            );
+        } else {
+            $logger->info('No ViewHelpers used.');
```
This should also be put to console via $output

NamelessCoder commented on Classes/Command/FluidTemplatePrecompileCommand.php:
```diff
@@ -0,0 +1,161 @@
+        $uncompilable = array_sum($uncompilable);
+
+        $outputHelper->section('General');
+        $outputHelper->text(vsprintf('Compiled %d extension(s) with %d template(s) in total.', [count($result), $templatesCount]));
```
A more appropriate error message would be `Compiled %d template(s) in %d extension(s)`.

NamelessCoder commented on Classes/Command/FluidTemplateProfileCommand.php:
```diff
@@ -0,0 +1,162 @@
+
+        // General information
+        $outputHelper->section('General');
+        $rows = [
```
This looks almost entirely like `$rows = [$result]` would do the same; except for squashing the sections which could be done afterwards. Since `$result['sections']` doesn't get used below, you could simply do:

```php
$result['sections'] = count($result['sections']);
$outputHelper->table(['Info', 'Value'], [$result]);
```

To replace lines 52-88.

NamelessCoder reviewed (changes requested):
Needs a bit more love, see comments.

General CGL-related comments:

* Use `!$var` instead of `false === $var` when it's a boolean. When it may be not-boolean, `$var === false`.
* Avoid redundant comments, for example single words that contain the same text as the only string argument used directly below.
* Remove the double line breaks.
* `sprintf` is preferred over `vsprintf`; there's a much longer explanation but the short version is: `sprintf` when your arguments for the formatted string are explicit and not dynamically built. Use `vsprintf` when your arguments may be dynamic (as in: you don't know the exact argument count and types).
* `PHP_EOL` is always preferred over `"\n"`.
* Only use double-quoted strings when you must; using `PHP_EOL` means you almost never have to double quote.
